### ISSUE TITLE
Make Taipy GUI search and run local on_init(state) functions

### ISSUE THREAD

**Issue by cheak1974** [Avaiga/taipy#1341]
### Description

At the time Taipy is just searching and running the global on_init(state) function which is not very handy, because you have to manually import and call an local on_init function in the local page code.

Furthermore the code is only run once and gets the state id "global".  So when Taipy starts the on_init function is called once and it makes absolutely no sense to forward the state object to the call of the local on_init function like mentioned in the docs, because it isn't the state of the new user session/connection but stays always the global state.

### Solution Proposed

I would propose to call a function on_init(state) in each local module which would give the user the possibility to start e.g. data aquisition threads when a new connection opens up. Very useful could also be a standard variable bound to the state object to query if the connection is still alive. This would be great for ending such data aquisition threads when the user disconnects.

### Impact of Solution

none

### Additional Context

none

### Acceptance Criteria

- [ ] Ensure new code is unit tested, and check code coverage is at least 90%.
- [ ] Create related issue in taipy-doc for documentation and Release Notes.
- [ ] Check if a new demo could be provided based on this, or if legacy demos could be benefit from it.
- [ ] Ensure any change is well documented.

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

**Comment by FlorianJacta** [comment]
How is it not called for each new state? It should. Otherwise, I like the idea but I am not sure how to decide which *on_init* should be called first if you have multiple ones. It should create some issues

**Comment by FredLL-Avaiga** [comment]
Would you care to share some reproduction code for the global state issue?

**Comment by FabienLelaquais** [comment]
Isn't this issue covered by Page Modules?

**Comment by dinhlongviolin1** [comment]
Hi @cheak1974,

Taipy `on_init` function is designed to run only once only and executed by the __main__ module. We have come up with another function that is available in Taipy GUI 4.0.0 which is called `on_page_load` that would be executed every time the page is loaded up on the browser. This function can also be used by other page modules so it should cover what you need. Please have a look at the example below.

### Example 1: single page demo
```python
# A simple temparature converter to demonstrate how to use the on_page_load

from taipy.gui import Gui

c_temp = 0
f_temp = 0

# reset the c_temp to 20 whenever the page is loaded/reloaded on the browser
# the f_temp will also be reseted due to the on_change function
def on_page_load(state):
    print("reseting c_temp to 20")
    state.c_temp = 20

def on_change(state, var_name, value):
    if var_name == 'c_temp':
        state.f_temp = value * 9/5 + 32

# The on_navigate function on paper can be used similarly to the on_page_load
# as it is called whenever the page is navigated/loaded/reloaded and can be used
# on multiple page modules. But it should be used only for directing navigation.
# The on_navigate function is called just before the on_page_load function
def on_navigate(state, page_name):
    print("navigating to page:", page_name)
    return page_name

page_md = """
# Temperature Converter

The temperature in Celcius: <|{c_temp}|number|>

The temperature in Fahrenheit: <|{f_temp:.1f}|>
"""

Gui(page_md).run()
```

### Example 2: multiple page modules
**temperature-converter-demo.py**
```python
from taipy.gui import Gui

from page_fahrenheit import page as page1

c_temp = 0
f_temp = 0

def on_page_load(state):
    state.c_temp = 20

def on_change(state, var_name, value):
    if var_name == 'c_temp':
        state.f_temp = value * 9/5 + 32

page_md = """
# Temperature Converter Celcius to Fahrenheit

The temperature in Celcius: <|{c_temp}|number|>

The temperature in Fahrenheit: <|{f_temp:.1f}|>
"""

pages = {
    "celcius": page_md,
    "fahrenheit": page1,
}

Gui("<|navbar|>", pages).run()
```
**page_fahrenheit.py**
```python
from taipy.gui import Markdown

c_temp = 0
f_temp = 0

def on_page_load(state):
    state.f_temp = 68

def on_change(state, var_name, value):
    if var_name == 'f_temp':
        state.c_temp = (value - 32) * 5/9

page_md = """
# Temperature Converter Fahrenheit to Celcius

The temperature in Fahrenheit: <|{f_temp}|number|>

The temperature in Celcius: <|{c_temp:.1f}|>
"""

page = Markdown(page_md)
```

I will proceed to close this issue. If you have any question/problem, you can reopen this issue or contact me directly through nguyen.dinh.long@taipy.io
